Plutarch/Rational.hs

```diff
@@ -192,48 +217,36 @@ instance PFractional PRational where
     phoistAcyclic $
       plam $ \x ->
         pmatch x $ \(PRational xn xd) ->
-          pfailOnZero # xn # pcon (PRational xd xn)
+          pcon $ PRational (pto xd) $ ptryPositive # xn
 
   x' #/ y' =
     phoistAcyclic
-      ( plam $ \x y ->
-          preduce
-            #$ pmatch x
-            $ \(PRational xn xd) ->
-              pmatch y $ \(PRational yn yd) ->
-                plet (xd * yn) $ \den ->
-                  pfailOnZero # den
-                    #$ pcon (PRational (xn * yd) den)
+      ( plam $ \x y -> unTermCont $ do
+          PRational xn xd <- tcont $ pmatch x
+          PRational yn yd <- tcont $ pmatch y
+          denm <- tcont . plet $ ptryPositive #$ pto xd * yn
+          pure $ preduce #$ pcon $ PRational (xn * pto yd) denm
       )
       # x'
       # y'
 
   pfromRational = phoistAcyclic $ plam id
 
-pfailOnZero :: Term s (PInteger :--> a :--> a)
-pfailOnZero = phoistAcyclic $
-  plam $ \n x ->
-    pif
-      (n #== 0)
-      (ptraceError "pfailOnZero: division by 0")
-      x
-
 preduce :: Term s (PRational :--> PRational)
 preduce = phoistAcyclic $
-  plam $ \x ->
-    pmatch x $ \(PRational xn xd) ->
-      plet (pgcd # xn # xd) $ \r ->
-        plet (signum xd) $ \s ->
-          pcon $ PRational (s * pdiv # xn # r) (s * pdiv # xd # r)
+  plam $ \x -> unTermCont $ do
+    PRational xn xd' <- tcont $ pmatch x
+    xd <- tcont . plet $ pto xd'
+    r <- tcont . plet $ pgcd # xn # xd
+    s <- tcont . plet . signum $ xd
```

Comment:
...signum? Not `psignum`?

Reply:
updated - are we preferring the Plutarch typeclass functions in general?

Reply:
Good question.